In [10]:
import numpy as np 
import pandas as pd
from sklearn.neural_network import MLPClassifier
import matplotlib.pyplot as plt

# STEP 1: Clean and Split Dataset

In [2]:
df = pd.read_csv("Titanic.csv")
df = df.drop(columns=['Unnamed: 0','name','ticket','fare','cabin','embarked','boat','body','home.dest','parch'])
df_new = df[['pclass','sex','age','sibsp','survived']]

pclass = {'1st': 1, '2nd': 2, '3rd': 3}
sex = {'female': 0, 'male': 1}

df_new['pclass'] = df_new['pclass'].map(pclass)
df_new['sex'] = df_new['sex'].map(sex)
df_new['age'] = df_new['age'].fillna(30)


shuffled = df_new.sample(frac=1)



split = int(.8*len(shuffled))
train = shuffled[:split]
test = shuffled[split:]

x_train = train.iloc[:,0:4]
y_train = train.iloc[:,4:5]

x_test = test.iloc[:,0:4]
y_test = test.iloc[:,4:5]

# Fit Neural Network #1 
## 2 hidden layers, 2 neurons each 

In [3]:
clf = MLPClassifier(solver = 'adam',alpha = .01,activation="logistic",hidden_layer_sizes=(2, 2),learning_rate_init=0.05,momentum=.9)

clf.fit(x_train,y_train)
clf.predict(x_test)

print("Score:",clf.score(x_test,y_test))

Score: 0.7748091603053435


/usr/local/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


# Check Perfomance of Neural Network #1 

In [4]:
df_pred = pd.DataFrame(data = clf.predict(x_test),columns=['pred'])
df_targ = pd.DataFrame(data = y_test)

df_pred.reset_index(drop=True, inplace=True)
df_targ.reset_index(drop=True, inplace=True)

df_fin = pd.concat( [df_pred, df_targ], axis=1) 



zeros = df_fin[(df_fin["pred"] == 0) & (df_fin["survived"] == 0)]
print(' fatalities correctly predicted:',len(zeros)/len(y_test)*100)
ones = df_fin[(df_fin["pred"] == 1) & (df_fin["survived"] == 1)]
print(' survivals correctly predicted:',len(ones)/len(y_test)*100)




 fatalities correctly predicted: 47.32824427480916
 survivals correctly predicted: 30.15267175572519


# Fit Neural Network #2
## 2 hidden layers, 20 neurons and 10 neurons  

In [5]:
clf2 = MLPClassifier(solver = 'adam',alpha = .01,activation="logistic",hidden_layer_sizes=(20, 10),learning_rate_init=0.05,momentum=.9,max_iter=35)


clf2.fit(x_train,y_train)
clf2.predict(x_test)

print("Score:",clf2.score(x_test,y_test))

Score: 0.8053435114503816


/usr/local/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (35) reached and the optimization hasn't converged yet.
  warnings.warn(


# Check Performance of Neural Network #2

In [13]:
df_pred2 = pd.DataFrame(data = clf2.predict(x_test),columns=['pred'])
df_targ2 = pd.DataFrame(data = y_test)

df_pred2.reset_index(drop=True, inplace=True)
df_targ2.reset_index(drop=True, inplace=True)

df_fin2 = pd.concat( [df_pred2, df_targ2], axis=1) 



zeros = df_fin2[(df_fin2["pred"] == 0) & (df_fin2["survived"] == 0)]
print(' fatalities correctly predicted:',len(zeros)/len(y_test)*100)
ones = df_fin2[(df_fin2["pred"] == 1) & (df_fin2["survived"] == 1)]
print(' survivals correctly predicted:',len(ones)/len(y_test)*100)

 fatalities correctly predicted: 48.473282442748086
 survivals correctly predicted: 32.06106870229007


# Accuracy of random forest 

## fatalities correctly predicted: 0.5190839694656488


## survivals correctly predicted: 0.29389312977099236

# Accuracy of neural network 

## fatalities correctly predicted: 48.473282442748086


## survivals correctly predicted: 32.06106870229007

the accuracy between the neural network and the random forest is very simnilar. The random forest was created in one iteration and the neural network was tested over one epoch. Neural networks and random forests are algorithms that are both suited for handling classification problems with non-linear data. I think that the data set was small enough that they were able to have similar performances. However, if the data set were larger we might find that the accuracy between the random forset and neural network might be more drastic. 